Mathlib/Tactic/TacticAnalysis/Declarations.lean

```diff
@@ -288,20 +289,105 @@ register_option linter.tacticAnalysis.tryAtEachStep.fraction : Nat := {
   defValue := 1
 }
 
-/-- Run a tactic at each proof step. -/
+/--
+Whether to show timing information in "tryAtEachStep" tactic analysis output.
+
+When true (default), messages include elapsed time like `(23ms)`.
+Set to false in tests to avoid non-deterministic output.
+-/
+register_option linter.tacticAnalysis.tryAtEachStep.showTiming : Bool := {
+  defValue := true
+}
+
+/-- Run a tactic at each proof step, with optional timing.
+
+Reports elapsed time in milliseconds for each successful replacement
+when `linter.tacticAnalysis.tryAtEachStep.showTiming` is true.
+-/
 def Mathlib.TacticAnalysis.tryAtEachStep (tac : Syntax → MVarId → CommandElabM (TSyntax `tactic)) : TacticAnalysis.Config where
   run seq := do
-    let fraction := linter.tacticAnalysis.tryAtEachStep.fraction.get (← getOptions)
+    let opts ← getOptions
+    let fraction := linter.tacticAnalysis.tryAtEachStep.fraction.get opts
+    let showTiming := linter.tacticAnalysis.tryAtEachStep.showTiming.get opts
     for i in seq do
       if let [goal] := i.tacI.goalsBefore then
         if (hash goal) % fraction = 0 then
           let tac ← tac i.tacI.stx goal
+          let startTime ← IO.monoMsNow
           let goalsAfter ← try
             i.runTacticCode goal tac
           catch _e =>
             pure [goal]
+          let elapsedMs := (← IO.monoMsNow) - startTime
           if goalsAfter.isEmpty then
-            logInfoAt i.tacI.stx m!"`{i.tacI.stx}` can be replaced with `{tac}`"
+            if showTiming then
+              logInfoAt i.tacI.stx m!"`{i.tacI.stx}` can be replaced with `{tac}` ({elapsedMs}ms)"
+            else
+              logInfoAt i.tacI.stx m!"`{i.tacI.stx}` can be replaced with `{tac}`"
+
+/-- Run a tactic (given as a string) at each proof step, with optional timing.
+
+`label` is the human-readable name shown in output (e.g., "grind").
+`tacticStr` is the tactic syntax as a string (e.g., "grind +suggestions").
+Tactic sequences like "simp; grind" are also supported.
+
+Reports elapsed time in milliseconds for each successful replacement
+when `linter.tacticAnalysis.tryAtEachStep.showTiming` is true.
+To limit tactic runtime, use `set_option maxHeartbeats N` in the build command.
+-/
+def Mathlib.TacticAnalysis.tryAtEachStepFromStrings
+    (label : String) (tacticStr : String) : TacticAnalysis.Config where
+  run seq := do
+    -- Parse using `tacticSeq.fn` directly since `tacticSeq` is not a parser category.
+    -- See https://leanprover.zulipchat.com/#narrow/channel/113488-general/topic/piggy.20back.20off.20of.20the.20lean4.20parser
+    let tacSeq ← ofExcept <|
+      Mathlib.GuardExceptions.captureException (← getEnv) Parser.Tactic.tacticSeq.fn tacticStr
```

Comment:
Ah, I gather that using `Parser.Tactic.tacticSeq` here allows the input to be multiple tactics chained together like `tac1; tac2`. I like this idea. I use `Parser.tacticParser` [in my tryAtEachStep](https://github.com/dwrensha/tryAtEachStep/blob/2f43558f26b9996451f41c72f58b94133b94f23d/tryAtEachStep.lean#L328-L338), which requires outer parentheses for chaining, like `(tac1; tac2)`.